File: src/main/kotlin/com/linchpino/ai/service/impl/RoadmapServiceImpl.java
package com.linchpino.ai.service.impl;

import com.linchpino.ai.model.AIServiceName;
import com.linchpino.ai.model.RequestDetail;
import com.linchpino.ai.model.Resume;
import com.linchpino.ai.service.AIService;
import com.linchpino.ai.service.RoadmapService;
import io.grpc.netty.shaded.io.netty.util.internal.ObjectUtil;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.ApplicationContext;
import org.springframework.stereotype.Service;

import java.io.File;

@Service
public class RoadmapServiceImpl implements RoadmapService {

    @Value("${spring.ai.default-service-provider}")
    private String aiServiceName;

    private final ApplicationContext applicationContext;
    private final ResumeService resumeService;

    public RoadmapServiceImpl(ApplicationContext applicationContext, ResumeService resumeService) {
        this.applicationContext = applicationContext;
        this.resumeService = resumeService;
    }

    public AIService loadAIService(AIServiceName serviceName) {
        return applicationContext.getBean(serviceName.getComponentName(), AIService.class);
    }

    public String getRoadmap(String target, File resumeFile) {
        ObjectUtil.checkNotNull(target, "Target level cannot be null");
        ObjectUtil.checkNotNull(resumeFile, "Resume file cannot be null");
        AIServiceName serviceName = AIServiceName.getComponentNameOrDefault(aiServiceName);
        Resume resume = resumeService.create(resumeFile);
        RequestDetail requestDetail = new RequestDetail(target, resumeService.save(resume));
        return loadAIService(serviceName).talkToAI(requestDetail);
    }

}

File: src/main/kotlin/com/linchpino/ai/service/impl/RoadmapServiceImpl.kt
package com.linchpino.ai.service.impl

import com.linchpino.ai.model.AIServiceName
import com.linchpino.ai.model.AIServiceName.Companion.getComponentNameOrDefault
import com.linchpino.ai.model.RequestDetail
import com.linchpino.ai.service.AIService
import com.linchpino.ai.service.RoadmapService
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.ApplicationContext
import org.springframework.stereotype.Service
import java.io.File

@Service
class RoadmapServiceImpl(private val applicationContext: ApplicationContext, private val resumeService: ResumeService) :
    RoadmapService {
    @Value("\${spring.ai.default-service-provider}")
    private val aiServiceName: String? = null

    fun loadAIService(serviceName: AIServiceName): AIService {
        return applicationContext.getBean(serviceName.componentName, AIService::class.java)
    }

    override fun getRoadmap(targetLevel: String, resumeFile: File): String? {
        val serviceName = getComponentNameOrDefault(aiServiceName)
        val resume = resumeService.create(resumeFile)
        val requestDetail = RequestDetail(targetLevel, resumeService.save(resume))
        return loadAIService(serviceName).talkToAI(requestDetail)
    }
}
